Python/immutability.c

```diff
@@ -0,0 +1,890 @@
+        PyErr_SetObject(PyExc_TypeError, error_msg);
+        return -1;
+    }
+
+    return 0;
+}
+
+
+int _PyImmutability_RegisterFreezable(PyTypeObject* tp)
+{
+    PyObject *ref;
+    int result;
+    struct _Py_immutability_state *state = get_immutable_state();
+    if(state == NULL){
+        PyErr_SetString(PyExc_RuntimeError, "Failed to initialize immutability state");
+        return -1;
+    }
+
+    ref = type_weakref(state, (PyObject*)tp);
+    if(ref == NULL){
+        return -1;
+    }
+
+    result = PySet_Add(state->freezable_types, ref);
+    Py_DECREF(ref);
+    return result;
+}
+
+// Perform a decref on an immutable object
+// returns true if the object should be deallocated.
+int _Py_DecRef_Immutable(PyObject *op)
+{
+    // Decrement the reference count of an immutable object without
+    // deallocating it.
+    assert(_Py_IsImmutable(op));
+
+#ifdef Py_GIL_DISABLED
+    // Put the clear code in DecRefShared.
+    _Py_DecRefShared(op);
+    return false;
+#else
+    // TODO(Immutable): This will need to be atomic.
+    op->ob_refcnt -= 1;
+    if (_Py_IMMUTABLE_FLAG_CLEAR(op->ob_refcnt) != 0)
+        // Context does not to dealloc this object.
+        return false;
+
+    assert(_Py_IMMUTABLE_FLAG_CLEAR(op->ob_refcnt) == 0);
+
+    _Py_CLEAR_IMMUTABLE(op);
+
+    return true;
+#endif
+}
+
+// Macro that jumps to error, if the expression `x` does not succeed.
+#define SUCCEEDS(x) { do { int r = (x); if (r != 0) goto error; } while (0); }
+
+int traverse_freeze(PyObject* obj, PyObject* dfs)
+{
+    if(is_c_wrapper(obj)) {
+        // C functions are not mutable
+        // Types are manually traversed
+        return 0;
+    }
+
+    // Function require some work to freeze, so we do not freeze the
+    // world as they mention globals and builtins.  This will shadow what they
+    // use, and then we can freeze the those components.
+    if(PyFunction_Check(obj)){
+        SUCCEEDS(shadow_function_globals(obj));
+    }
+
+    if(PyType_Check(obj)){
+        // TODO(Immutable): mjp: Special case for types not sure if required. We should review.
+        PyTypeObject* type = (PyTypeObject*)obj;
+
+        SUCCEEDS(freeze_visit(type->tp_dict, dfs));
+        SUCCEEDS(freeze_visit(type->tp_mro, dfs));
+        // We need to freeze the tuple object, even though the types
+        // within will have been frozen already.
+        SUCCEEDS(freeze_visit(type->tp_bases, dfs));
+    }
+    else
+    {
+        traverseproc traverse = Py_TYPE(obj)->tp_traverse;
+        if(traverse != NULL){
+            SUCCEEDS(traverse(obj, (visitproc)freeze_visit, dfs));
+        }
+    }
+
+    // The default tp_traverse will not visit the type object if it is
+    // not heap allocated, so we need to do that manually here to freeze
+    // the statically allocated types that are reachable.
+    if (!(Py_TYPE(obj)->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
+        SUCCEEDS(freeze_visit(_PyObject_CAST(Py_TYPE(obj)), dfs));
+    }
```

Comment:
I don't believe that `tp_traverse` visits the type, regardless of if it's heap allocated or not. I checked a few `tp_traverse` implementations and non visited the type. This extra rule also caused a violation of the immutability invariant. After freezing an object of type `_ctypes.PyCArrayType` the object itself was immutable, but could still reach the mutable `_ctypes.PyCArrayType` type.

Reply:
So I have issues with reference counting bugs without this in the SCC code.  But we can back it out, and see if the shape has changed in 3.15.  It might be that the type field is only RCed if it is a HEAP_TYPE, and this fixed that, but broke the invariant.

Reply:
I'm guessing that these might be two different issues:
1. Immutability should propagate to type objects, regardless if they're head of static objects
2. The SCC as I last heard about it was using the GC header to store information. So the SCC algorithm should be limited to heap types.

Does this guess sound plausable?

---

I also found special handling for heap types in `type_traverse`.
https://github.com/mjp41/cpython/blob/3a215b7810c1134e3edc3a23b7a144151a454f24/Objects/typeobject.c#L6888-L6895

This prevents us from calling `tp_traverse` on static type objects, but we can still call the `visit` function with the static type being the target.

In #59 I removed this restriction and everything seems to be working™️